src/snowflake/connector/connection.py

```diff
@@ -314,6 +314,25 @@ class TypeAndBinding(NamedTuple):
     binding: str | None
 
 
+# SNOW-1669128 previously, the close_at_exit function for the SnowflakeConnection object
+# was registered with the `atexit` module during initialization of a SnowflakeConnection object.
+# This caused the Snowpark Session object's atexit function to be registered before the
+# SnowflakeConnection's object when the Session was made without an existing SnowflakeConnection
+# object, because the SnowflakeConnection object's atexit was only registered after it was initialized
+# whereas the Snowpark Session object's atexit was registered when the file was imported. During the
+# atexit for the Snowpark Session, it attempts to send telemetry regarding cleanup operations, but since
+# it's connection is already closed, it fails to send the telemetry. Moving the registration of the atexit
+# function here means that it will happen when the SnowflakeConnection object is imported - and so the
+# Snowpark Session's atexit function will be registered after, and the order will be correct.
+def _close_connection_at_exit(snowflake_connection: SnowflakeConnection):
```

Comment:
Where is this parameter passed in from?

Reply:
Good catch, thank you! I've added a variable to track active connections - would this work, or would you suggest another workaround (e.g. perhaps by making a class method to register the atexit function in the SnowflakeConnection object, and calling it from the Snowpark Session object before registering its own atexit function)?